File: py/behavioral/strategy.py
# Strategy injects interchangeable algorithms behind the same calling shape.
# Modern Python note: first-class functions are usually the simplest Strategy. Use
# classes when a strategy carries state, dependencies, or lifecycle.
def standard_shipping(order_total):
    return 5


def free_over_50_shipping(order_total):
    return 0 if order_total >= 50 else 5


class Checkout:
    def __init__(self, shipping_strategy):
        self.shipping_strategy = shipping_strategy

    def total(self, subtotal):
        return subtotal + self.shipping_strategy(subtotal)


if __name__ == "__main__":
    checkout = Checkout(free_over_50_shipping)
    print(checkout.total(60))

File: csharp/behavioral/strategy.cs
namespace Gof.ModernCSharp.Behavioral;

// Strategy injects interchangeable algorithms behind the same calling shape.
// Modern C# note: delegates, lambdas, and Func<T> are usually the simplest form.
// Classes help when strategies carry state, dependencies, or lifecycle.
public static class Strategy
{
    static int FreeOver50Shipping(int orderTotal) => orderTotal >= 50 ? 0 : 5;

    sealed class Checkout(Func<int, int> shippingStrategy)
    {
        public int Total(int subtotal) => subtotal + shippingStrategy(subtotal);
    }

    public static string Run() => new Checkout(FreeOver50Shipping).Total(60).ToString();
}
